tools/autograd/derivatives.yaml

```diff
@@ -695,13 +695,13 @@
   self: hardshrink_backward(grad, self, lambd)
 
 - name: hardtanh(Tensor self, Scalar min_val, Scalar max_val)
-  self: hardtanh_backward(grad, self, min_val, max_val)
+  self: hardtanh_backward(grad, output, min_val, max_val)
 
 - name: hardtanh_(Tensor self, Scalar min_val, Scalar max_val)
   self: hardtanh_backward(grad, output, min_val, max_val)
 
 - name: leaky_relu(Tensor self, Scalar negative_slope)
-  self: leaky_relu_backward(grad, self, negative_slope)
+  self: leaky_relu_backward(grad, output, negative_slope)
```

Comment:
Does this only expand to calling `at::leaky_relu_backward(grad, output, negative_slope)`, or do we still have some more complicated rules for NN? It's weird that swapping input for output is still handled correctly O_o

Reply:
I had the same doubt regarding relu.

Using `output` in non-inplace `threshold` led to a failed test, while tests were passing for `leaky_relu`, `rrelu` and `hardtanh`, so I kept using `output`. However, the real fix brought in by the PR is on `get_variable_type.py`: there's no need to use `output` in non-inline now.

Reply:
We can go back to using `self` - maybe tests are just insufficient to demonstrate the problem on those activation functions. Again, the `memcpy` issue for `inplace` is solved anyway.

Reply:
Ok, looking into this more closely: you can swap input and output for `leaky_relu` and `rrelu`.
The derivative is a constant that depends on the sign of the input, which is equal to the sign of the output.

For `hardtanh`: the derivative is 0 for `x<-1` and `x>1` and 1 otherwise. Input and output are the same in the `-1 <= x <= 1` interval, so we're good. For `x<-1` and `x>1` the output is `y = -1` and `y = 1`, respectively, which (I looked in THNN) returns 0. So they're effectively the same.

Reply:
My concerns are not so much about fixing the memory usage issues, but more about correctness. We can't be passing outputs to backwards that expect inputs and vice versa.

BTW this is not true for leaky relu if `negative_slope` is positive - output will be `>= 0`.

Reply:
We definitely **should** go back to using `self` if that's what real arguments are. If we ever want to pass outputs instead of `self`, we should change the THNN interface to make sure we don't break stuff when refactoring other parts of the codebase.

Reply:
You're right, I'll submit a PR now.

Reply:
The in-place activation functions are precisely the ones where you can swap input and output. That's how THNN is written. `negative_slope <= 0` was never acceptable. We should probably add a check for that to LeakyRELU.c/cu.

Reply:
In general, if you can save input or output without any perf implication, it's a little nicer to save the input because subsequent calls are more likely to modify the output than the input.

For example, if you have separate derivative definitions for `relu` and `relu_` save `self` for `relu` and `output` for `relu_`.

Reply:
We can do that with the current codegen, use `self` for non-inplace and `output` for inplace.